File: src/converter/plugins/AnnotationPlugin.ts
import {ConverterPlugin} from "../plugin.js";
import ts, {Node} from "typescript";
import {ConverterContext} from "../context.js";
import {Render} from "../render.js";
import {Annotation, AnnotationContext} from "../annotation.js";
import {GeneratedFile} from "../generated.js";

export const annotationServiceKey = Symbol()

export class AnnotationService {
    constructor(readonly annotations: Annotation[]) {
    }

    resolveAnonymousAnnotations(node: Node, context: ConverterContext): string[] {
        return this.internalResolveAnnotations(node, true, context)
    }

    resolveAnnotations(node: Node, context: ConverterContext): string[] {
        return this.internalResolveAnnotations(node, false, context)
    }

    private internalResolveAnnotations(node: Node, isAnonymousDeclaration: boolean, context: ConverterContext) {
        const annotationContext: AnnotationContext = {
            ...context,
            isAnonymousDeclaration,
        }

        const annotations: string[] = []

        for (const annotation of this.annotations) {
            const result = annotation(node, annotationContext)

            if (result !== null) annotations.push(result)
        }

        return annotations
    }
}

export class AnnotationPlugin implements ConverterPlugin {
    private readonly annotationService: AnnotationService

    constructor(annotations: Annotation[]) {
        this.annotationService = new AnnotationService(annotations)
    }

    generate(): GeneratedFile[] {
        return [];
    }

    render(node: ts.Node, context: ConverterContext, next: Render): string | null {
        const annotations = this.annotationService.resolveAnnotations(node, context)

        if (annotations.length > 0) {
            return `
${annotations.join("\n")}
${next(node)}
            `.trim();
        }

        return null
    }

    setup(context: ConverterContext): void {
        context.registerService(annotationServiceKey, this.annotationService)
    }

    traverse(node: ts.Node, context: ConverterContext): void {
    }
}

File: gradle-plugin/karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/AnnotationPlugin.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.*
import js.array.ReadonlyArray
import js.symbol.Symbol
import typescript.Node

val annotationServiceKey = Symbol()

class AnnotationService(private val annotations: ReadonlyArray<Annotation<Node>>) {
    fun resolveAnonymousAnnotations(node: Node, context: ConverterContext): ReadonlyArray<String> {
        return internalResolveAnnotations(node, true, context)
    }

    fun resolveAnnotations(node: Node, context: ConverterContext): ReadonlyArray<String> {
        return internalResolveAnnotations(node, false, context)
    }

    private fun internalResolveAnnotations(
        node: Node,
        isAnonymousDeclaration: Boolean,
        context: ConverterContext,
    ): ReadonlyArray<String> {
        val annotationContext = object : AnnotationContext, ConverterContext by context {
            override val isAnonymousDeclaration = isAnonymousDeclaration
        }

        val annotations = mutableListOf<String>()

        for (annotation in this.annotations) {
            val result = annotation(node, annotationContext)

            if (result != null) annotations += result
        }

        return annotations.toTypedArray()
    }
}

class AnnotationPlugin(annotations: ReadonlyArray<Annotation<Node>>) : ConverterPlugin<Node> {
    private val annotationService = AnnotationService(annotations)

    override fun setup(context: ConverterContext) {
        context.registerService(annotationServiceKey, annotationService)
    }

    override fun traverse(node: Node, context: ConverterContext) = Unit

    override fun generate(context: ConverterContext, render: Render<Node>) = emptyArray<GeneratedFile>()

    override fun render(node: Node, context: ConverterContext, next: Render<Node>): String? {
        val annotations = annotationService.resolveAnnotations(node, context)

        if (annotations.isNotEmpty()) {
            return """
${annotations.joinToString(separator = "\n")}
${next(node)}
            """.trim();
        }

        return null
    }


}
